src/jobserver-posix.cc

```diff
@@ -0,0 +1,164 @@
+  else if (!jobserver_name_.empty())
+    jobserver_closed_ = true;
+
+  // Signal that we have initialized but do not have a token yet.
+  if (Enabled())
+    token_count_ = -1;
+}
+
+void PosixJobserverClient::Parse() {
+  // Return early if no makeflags are passed in the environment.
+  char* makeflags = std::getenv("MAKEFLAGS");
+  if (makeflags == nullptr || strlen(makeflags) == 0)
+    return;
+
+  std::string::size_type flag_char = 0;
+  std::string flag;
+  std::vector<std::string> flags;
+
+  // Tokenize string to characters in flag, then words in flags.
+  while (flag_char < strlen(makeflags)) {
+    while (flag_char < strlen(makeflags) &&
+           !isblank(static_cast<unsigned char>(makeflags[flag_char]))) {
+      flag.push_back(static_cast<unsigned char>(makeflags[flag_char]));
+      flag_char++;
+    }
+
+    if (!flag.empty())
+      flags.push_back(flag);
+
+    flag.clear();
+    flag_char++;
+  }
+
+  // --jobserver-auth=<val>
+  for (size_t n = 0; n < flags.size(); n++)
+    if (flags[n].find(kAuthKey) == 0)
+      flag = flags[n].substr(strlen(kAuthKey));
+
+  // --jobserver-fds=<val>
+  if (flag.empty())
+    for (size_t n = 0; n < flags.size(); n++)
+      if (flags[n].find(kFdsKey) == 0)
+        flag = flags[n].substr(strlen(kFdsKey));
+
+  // -j 1
+  if (flag.empty())
+    for (size_t n = 0; n < flags.size(); n++)
+      if (flags[n].find("-j") == 0)
+        jobserver_closed_ = true;
+
+  // Check for fifo pipe.
+  if (flag.find(kFifoKey) == 0)
+    jobserver_fifo_ = true;
+
+  jobserver_name_.assign(flag);
+}
+
+bool PosixJobserverClient::Enabled() const {
+  return (rfd_ >= 0 && wfd_ >= 0) || jobserver_closed_;
+}
+
+unsigned char PosixJobserverClient::Acquire() {
+  unsigned char token = '\0';
+
+  // The first token is implicitly handed to a process.
+  // Fallback to non-parallel if jobserver-capable parent has no pipe.
+  if (token_count_ <= 0 || jobserver_closed_) {
+    token_count_ = 1;
+    return token;
+  }
+
+  int ret = read(rfd_, &token, 1);
+  if (ret < 0 && errno != EAGAIN && errno != EWOULDBLOCK) {
+    jobserver_closed_ = true;
+    if (!jobserver_fifo_)
+      Warning("pipe closed: %d (mark the command as recursive)", rfd_);
+    else
+      Fatal("failed to read from jobserver: %d: %s", rfd_, strerror(errno));
+  }
+
+  if (ret > 0)
+    token_count_++;
+
+  return token;
+}
+
+void PosixJobserverClient::Release(unsigned char* token) {
+  if (token_count_ < 0)
+    token_count_ = 0;
+  if (token_count_ > 0)
+    token_count_--;
+
+  // The first token is implicitly handed to a process.
+  // Writing is not possible if the pipe is closed.
+  if (*token == '\0' || jobserver_closed_)
+    return;
+
+  int ret = write(wfd_, token, 1);
+  if (ret != 1) {
+    Fatal("failed to write to jobserver: %d: %s", wfd_, strerror(errno));
+  }
+
+  *token = '\0';
+}
+
+void PosixJobserverClient::Clear(std::vector<Edge*> edges) {
+  for (size_t n = 0; n < edges.size(); n++)
+    Release(&edges[n]->job_token_);
```

Comment:
Can you avoid puting this code here to avoid un-necessary coupling (i.e. the JobServer class doesn't need to know anything about the Edge class). There is also no reason to pass a vector by value here, which will perform a copy, use a const reference instead, and consider using for-loop constructs, e.g.:

```
void ClearJobTokens(const std::vector<Edge*> edges, JobServer* jobserver) {
  for (const Edge* edge : edges)
    jobserver->Release(&edge->job_token_);
}
```